File: resolve-conflicts.js
const fs = require('fs');
const path = require('path');
const glob = require('glob');

// Pattern to resolve test file conflicts - keep the version without "const vi" declaration
function resolveTestFileConflicts(content) {
  // Remove the vi declaration conflict - prefer the version without it
  content = content.replace(
    /<<<<<<< HEAD\n\s*const vi: typeof import\('vitest'\)\.vi\n=======\n>>>>>>>[^\n]+/g,
    ''
  );
  
  // For other conflicts in test files, prefer the incoming branch changes
  content = content.replace(
    /<<<<<<< HEAD\n([^=]*)=======\n([^>]*)>>>>>>>[^\n]+/g,
    (match, ours, theirs) => theirs
  );
  
  return content;
}

// Files to process
const testFiles = [
  'src/main/__tests__/*.test.ts',
  'src/preload/__tests__/*.test.ts', 
  'src/renderer/src/components/error/__tests__/*.test.tsx',
  'src/renderer/src/components/task/__tests__/*.test.tsx',
  'src/renderer/src/store/__tests__/*.test.ts',
  'tests/setup/preload.setup.ts'
];

// Process each test file
testFiles.forEach(pattern => {
  const files = glob.sync(pattern, { cwd: __dirname });
  
  files.forEach(file => {
    const filePath = path.join(__dirname, file);
    try {
      let content = fs.readFileSync(filePath, 'utf8');
      if (content.includes('<<<<<<<')) {
        content = resolveTestFileConflicts(content);
        fs.writeFileSync(filePath, content);
        console.log(`Resolved conflicts in: ${file}`);
      }
    } catch (err) {
      console.error(`Error processing ${file}:`, err.message);
    }
  });
});

// Handle specific files
const specificFiles = {
  'startup.sh': (content) => {
    // For startup.sh, prefer the incoming changes
    return content.replace(/<<<<<<< HEAD\n([^=]*)=======\n([^>]*)>>>>>>>[^\n]+/g, (m, o, t) => t);
  },
  'test-results/results.json': (content) => {
    // For JSON, prefer the incoming valid JSON
    return content.replace(/<<<<<<< HEAD\n([^=]*)=======\n([^>]*)>>>>>>>[^\n]+/g, (m, o, t) => t);
  },
  'src/renderer/src/hooks/useErrorBoundary.ts': (content) => {
    // For useErrorBoundary, prefer incoming changes
    return content.replace(/<<<<<<< HEAD\n([^=]*)=======\n([^>]*)>>>>>>>[^\n]+/g, (m, o, t) => t);
  },
  'wt_tasks/progress.md': (content) => {
    // For progress.md, prefer incoming changes
    return content.replace(/<<<<<<< HEAD\n([^=]*)=======\n([^>]*)>>>>>>>[^\n]+/g, (m, o, t) => t);
  }
};

// Process specific files
Object.entries(specificFiles).forEach(([file, resolver]) => {
  const filePath = path.join(__dirname, file);
  try {
    let content = fs.readFileSync(filePath, 'utf8');
    if (content.includes('<<<<<<<')) {
      content = resolver(content);
      fs.writeFileSync(filePath, content);
      console.log(`Resolved conflicts in: ${file}`);
    }
  } catch (err) {
    console.error(`Error processing ${file}:`, err.message);
  }
});

File: resolve_conflicts.py
#!/usr/bin/env python3
import re
import sys
import os

def resolve_conflicts(file_path):
    """Resolve merge conflicts in test files."""
    with open(file_path, 'r') as f:
        content = f.read()
    
    # Pattern to match the vi declaration conflict
    vi_conflict_pattern = r'<<<<<<< HEAD\n\s*const vi: typeof import\(\'vitest\'\)\.vi\n=======\n>>>>>>> test/test-failure-analysis'
    
    # Remove the vi declaration conflict (keep the version without it)
    content = re.sub(vi_conflict_pattern, '', content)
    
    # For remaining conflicts, generally prefer the incoming branch (test/test-failure-analysis)
    # This pattern captures conflicts and keeps the content after =======
    general_conflict_pattern = r'<<<<<<< HEAD\n(.*?)\n=======\n(.*?)\n>>>>>>> test/test-failure-analysis'
    
    def keep_theirs(match):
        return match.group(2)
    
    content = re.sub(general_conflict_pattern, keep_theirs, content, flags=re.DOTALL)
    
    # Write back the resolved content
    with open(file_path, 'w') as f:
        f.write(content)
    
    print(f"Resolved conflicts in: {file_path}")

# List of files with conflicts
conflict_files = [
    "src/main/__tests__/auto-updater-differential.test.ts",
    "src/main/__tests__/auto-updater-integration.test.ts",
    "src/main/__tests__/auto-updater-security.test.ts",
    "src/main/__tests__/auto-updater.test.ts",
    "src/main/__tests__/baseline.test.ts",
    "src/main/__tests__/cross-process-communication.test.ts",
    "src/main/__tests__/ipc-security.test.ts",
    "src/main/__tests__/lifecycle.test.ts",
    "src/main/__tests__/memory-leak-detection.test.ts",
    "src/main/__tests__/memory-leaks.test.ts",
    "src/main/__tests__/mocked-apis.test.ts",
    "src/main/__tests__/mocked-electron-apis.test.ts",
    "src/preload/__tests__/preload-security.test.ts",
    "src/renderer/src/components/error/__tests__/ErrorBoundary.test.tsx",
    "src/renderer/src/components/task/__tests__/TaskCard.test.tsx",
]

# Process each file
for file_path in conflict_files:
    if os.path.exists(file_path):
        try:
            resolve_conflicts(file_path)
        except Exception as e:
            print(f"Error processing {file_path}: {e}")
